src/org/zaproxy/zap/extension/custompages/ExtensionCustomPages.java

```diff
@@ -0,0 +1,232 @@
+import org.zaproxy.zap.model.ContextDataFactory;
+import org.zaproxy.zap.view.AbstractContextPropertiesPanel;
+import org.zaproxy.zap.view.ContextPanelFactory;
+
+/**
+ * A ZAP extension which implements Custom Pages, which
+ * allows you to define custom pages which may be used in various
+ * ZAP components.
+ */
+public class ExtensionCustomPages extends ExtensionAdaptor
+		implements ContextPanelFactory, ContextDataFactory, SessionChangedListener {
+
+	// The name is public so that other extensions can access it
+	public static final String NAME = "ExtensionCustomPages";
+	
+	private static final Logger LOGGER = Logger.getLogger(ExtensionCustomPages.class);
+	
+	// The i18n prefix
+	protected static final String PREFIX = "custompages";
+
+	public static final String CONTEXT_CONFIG_CUSTOM_PAGES = Context.CONTEXT_CONFIG + ".custompages";
+	public static final String CONTEXT_CONFIG_CUSTOM_PAGE = CONTEXT_CONFIG_CUSTOM_PAGES + ".page";
+
+	private static final int TYPE_CUSTOM_PAGE = 600;
+	
+	/** The Custom Page panels, mapped to each context. */
+	private Map<Integer, ContextCustomPagePanel> customPagePanelsMap = new HashMap<>();
+
+	/** The context managers, mapped to each context. */
+	private Map<Integer, ContextCustomPageManager> contextManagers = new HashMap<>();
+
+	public ExtensionCustomPages() {
+		super();
+		initialize();
+	}
+
+	private void initialize() {
+		this.setName(NAME);
+	}
+	
+	@Override
+	public String getUIName() {
+		return Constant.messages.getString("custompages.name");
+	}
+
+	@Override
+	public void hook(ExtensionHook extensionHook) {
+		super.hook(extensionHook);
+
+		extensionHook.addSessionListener(this);
+
+		// Register this as a context data factory
+		Model.getSingleton().addContextDataFactory(this);
+
+		if (getView() != null) {
+			// Factory for generating Session Context custompage panels
+			getView().addContextPanelFactory(this);
+		}
+
+	}
+
+	@Override
+	public String getAuthor() {
+		return Constant.ZAP_TEAM;
+	}
+
+	@Override
+	public String getDescription() {
+		return Constant.messages.getString(PREFIX + ".desc");
+	}
+
+	@Override
+	public URL getURL() {
+		try {
+			return new URL(Constant.ZAP_EXTENSIONS_PAGE);
+		} catch (MalformedURLException e) {
+			return null;
+		}
+	}
+
+	/**
+	 * Gets the context custom page manager for a given context.
+	 * 
+	 * @param contextId the context id
+	 * 
+	 * @return the context custom page manager
+	 */
+	public ContextCustomPageManager getContextCustomPageManager(int contextId) {
+		ContextCustomPageManager manager = contextManagers.get(contextId);
+		if (manager == null) {
+			manager = new ContextCustomPageManager(contextId);
+			contextManagers.put(contextId, manager);
+		}
+		return manager;
+	}
+
+	@Override
+	public void loadContextData(Session session, Context context) {
+		try {
+			List<String> encodedCustomPages = session.getContextDataStrings(context.getIndex(), TYPE_CUSTOM_PAGE);
+			for (String e : encodedCustomPages) {

```

Comment:
Don`t know, but `persistContextData()` has here a null check

Reply:
Looking at other extensions it seems this is a pattern. I guess if loadContextData is called then the Context isn't (can't be) null?

Reply:
Correct.

Reply:
right, thats fine